Avoid fatal error on ErrProcessDone

subpop opened the pull request:
When stopping a process, there is a chance the OS has already stopped the process. In that case, `Process.Kill` returns `os.ErrProcessDone`. This should not be considered a fatal error. This PR handles that case and does not return an error from `stopProcess` if the underlying `Process.Kill` call returns `os.ErrProcessDone`.

There are steps to reproduce this issue in #96, as well as [CCT-1066](https://issues.redhat.com/browse/CCT-1066).

I reproduced it by following steps 1 through 5 in CCT-1066, and then augmenting the running version of rhcd with one from the repository. There are steps in a comment in CCT-1066 that describe this process.

```
sudo go run -ldflags '-X github.com/redhatinsights/yggdrasil.PrefixDir=/usr -X github.com/redhatinsights/yggdrasil.LocalstateDir=/var -X github.com/redhatinsights/yggdrasil.ShortName=rhc -X github.com/redhatinsights/yggdrasil.LongName=rhc' ./cmd/yggd --config /etc/rhc/config.toml --log-level trace --topic-prefix redhat/insights
```

Press Ctrl-C to quit yggd. A message should appear in the console:

```
^Ccannot stop workers: cannot stop worker: cannot stop worker: cannot stop process os: process already finished
```

A PID file should exist at the location `/var/run/rhc/workers/rhc-worker-playbook.worker.pid`. It is the presence of this file, despite the process already being killed, that causes the next invocation of yggd from succeeding. Run the same command above again:
```
sudo go run -ldflags '-X github.com/redhatinsights/yggdrasil.PrefixDir=/usr -X github.com/redhatinsights/yggdrasil.LocalstateDir=/var -X github.com/redhatinsights/yggdrasil.ShortName=rhc -X github.com/redhatinsights/yggdrasil.LongName=rhc' ./cmd/yggd --config /etc/rhc/config.toml --log-level trace --topic-prefix redhat/insights
```

This time, it errors immediately:
```
[rhcd] 2025/02/05 14:23:54 /Users/link/Projects/yggdrasil/cmd/yggd/main.go:197: starting rhcd version
[rhcd] 2025/02/05 14:23:54 /Users/link/Projects/yggdrasil/cmd/yggd/main.go:199: attempting to kill any orphaned workers
cannot stop workers: cannot stop worker: cannot stop worker: cannot stop process os: process already finished
exit status 1
```

Checkout this branch and repeat the steps above. The second time `yggd` is run, the process should proceed instead of exiting with an error.

---
Card ID: [CCT-1066](https://issues.redhat.com/browse/CCT-1066)

jirihnidek reviewed (changes requested):
Overall, I like the fix. I have only small request.

subpop commented on cmd/yggd/exec.go:
```diff
@@ -81,7 +81,9 @@ func stopProcess(pid int) error {
-		return fmt.Errorf("cannot stop process %v", err)
+		if err != os.ErrProcessDone {
+			return fmt.Errorf("cannot stop process %v", err)
+		}
```
That's what happens here. If the return value of `process.Kill()` is not nil, it
contains an error. Inside the error-handling block, the error is checked; if it
is a known "ProcessDone" error, then exit the error-handling block and proceed
with the function. If the error is not "ErrProcessDone", wrap the error and
return it. Is it necessary to log it as well? The other functions in this file
don't log & return errors; they just return errors, leaving the logging of it up
to the caller.

jirihnidek replied:
I mean, when the process was already stopped by OS, shouldn't we print some log message in else statement. Something like this:

```go
if err != os.ErrProcessDone {
    return fmt.Errorf("cannot stop process %v", err)
} else {
    log.Debugf("cannot stop process: %v, it has been already terminated by OS", pid)
}
```

jirihnidek replied:
I see another risk. When you read worker the pid file of the worker, but the worker was already terminated by `kill -9`, then some other process (not related to yggdrasil at all) could be started with the same PID as killed worker... Then yggdrasil could kill some process (it is run using root user). The probability of such situation is low, but we should not blindly try to kill some process. Shouldn't we check first that the process really belong to the given worker (e.g. parsing `/proc/{pid}/cmdline)`?

jirihnidek reviewed (changes requested):
Oh, the PR contains more and more commits. :-) I have another request.

subpop commented on cmd/yggd/worker.go:
```diff
@@ -192,6 +192,18 @@ func stopWorker(name string) error {
+	if err != nil {
+		return fmt.Errorf("cannot read file /proc/%v/cmdline: %v", pid, err)
+	}
+	if filepath.Base(string(procCmdline)) != name {
```
I'm not sure this situation is possible. `yggd` is particular about the conventions required to start a worker. The file must be executable, must exist in `/usr/libexec/rhc` and must end with the string "worker". Because of these requirements, I don't believe it to be possible for a worker to be started and managed by `yggd` that has an interpreter the `cmdline` file.

However, a substring search instead of the base name check will also work, and does cover the case you described, so I'm happy to change it.

subpop commented:
> Oh, the PR contains more and more commits.

These rider are necessary. One drops Fedora from the build targets so that the packit checks will succeed. The other enables RPM builds so that packit can actually build RPMs for the 0.2 branch that are representative of the downstream rhc RPMs.

Archana-PandeyM commented:
I tried testing this PR on RHEL9.6 I could see the issue still exist :- 

[root@dell-per6525-02 ~]# rpm -qa | grep rhc
rhc-worker-playbook-0.1.10-1.el9.x86_64
rhc-0.2.6-0.6.git.5ba0073.el9.x86_64
[root@dell-per6525-02 ~]
[root@dell-per6525-02 ~]# rhc connect --username ***** --password ******
Connecting dell-***************** to Red Hat.
This might take a few seconds.

● Connected to Red Hat Subscription Management
● Connected to Red Hat Insights
● Activated the rhc daemon            <<<<<<<<<<<<<< why this message is changed?
● Enabled console.redhat.com services: remediations, compliance, remote configuration, insights

Successfully connected to Red Hat!

Manage your connected systems: https://red.ht/connector

STEP      DURATION  
rhsm      14.037s   
insights  37.214s   
rhc       24ms      
[root@dell-per6525-02 ~]# 
[root@dell-per6525-02 ~]# systemctl status rhcd
× rhcd.service - rhc daemon
     Loaded: loaded (/usr/lib/systemd/system/rhcd.service; disabled; preset: disabled)
    Drop-In: /etc/systemd/system/rhcd.service.d
             └─proxy.conf
     Active: failed (Result: exit-code) since Tue 2025-03-11 07:08:53 EDT; 5min ago
   Duration: 8ms
       Docs: https://github.com/redhatinsights/yggdrasil
    Process: 29558 ExecStart=/usr/sbin/rhcd (code=exited, status=1/FAILURE)
   Main PID: 29558 (code=exited, status=1/FAILURE)
        CPU: 9ms

Mar 11 07:08:53 dell-per6525-02.rhts.eng.pek2.redhat.com systemd[1]: Started rhc daemon.
Mar 11 07:08:53 dell-per6525-02.rhts.eng.pek2.redhat.com rhcd[29558]: [rhcd] 2025/03/11 07:08:53 /builddir/build/BUILD/rhc/yggdrasil-0.2.6/cmd/yggd/main.go:197: starting rhcd version 0.2.6
Mar 11 07:08:53 dell-per6525-02.rhts.eng.pek2.redhat.com rhcd[29558]: [rhcd] 2025/03/11 07:08:53 /builddir/build/BUILD/rhc/yggdrasil-0.2.6/cmd/yggd/main.go:199: attempting to kill any orphaned workers
Mar 11 07:08:53 dell-per6525-02.rhts.eng.pek2.redhat.com rhcd[29558]: cannot stop workers: cannot stop worker: cannot read file /proc/27380/cmdline: open /proc/27380/cmdline: no such file or directory
Mar 11 07:08:53 dell-per6525-02.rhts.eng.pek2.redhat.com systemd[1]: rhcd.service: Main process exited, code=exited, status=1/FAILURE
Mar 11 07:08:53 dell-per6525-02.rhts.eng.pek2.redhat.com systemd[1]: rhcd.service: Failed with result 'exit-code'.
[root@dell-per6525-02 ~]#

jirihnidek commented on cmd/yggd/worker.go:
```diff
@@ -192,6 +192,18 @@ func stopWorker(name string) error {
 
+	procCmdline, err := os.ReadFile(fmt.Sprintf("/proc/%v/cmdline", pid))
+	if err != nil {
+		return fmt.Errorf("cannot read file /proc/%v/cmdline: %v", pid, err)
```
@Archana-PandeyM tested that returning error is not the good idea. Why? When `workers/worker_name.pid` exists and it contains PID of non-existent process, then rhcd will not be able to start. We should only print some warning and delete `workers/worker_name.pid` file.

subpop replied:
Hmm. I wrote this error check under the assumption of other failures related to
reading the proc file (permissions, etc.) but you're right. If /proc/$PID does
not exist, the process is already stopped. Maybe we should check for the
existence of `/proc/$PID` explicitly, first?

jirihnidek commented on cmd/yggd/worker.go:
```diff
@@ -192,6 +192,18 @@ func stopWorker(name string) error {
+		return fmt.Errorf("cannot read file /proc/%v/cmdline: %v", pid, err)
+	}
+	if strings.Contains(string(procCmdline), name) {
+		return fmt.Errorf(
```
This also should not be fatal error. We should print some warning and delete obsolete pid file.

subpop replied:
Isn't this condition incorrect too? We should be returning from this function
early if procCmdline does *not* contain the worker name.

subpop commented:
Thank you for the review. I took a look over the `stopWorker` function in general and rewrote some of it. I moved the PID file removal into a defer block, so that it will always be removed at the end of the function, whether control makes it all the way through, or aborts early. I also added an explicit check for `/proc/$PID`; if that directory is absent, the function returns early.

Archana-PandeyM commented:
Thank you for the changes Link! Fix is working now-
I tried killing the PID of worker manually and rhcd was able to register the worker back.
Following logs appeared and registration succeeded and
```

Mar 12 05:11:05 dell-per6525-02.rhts.eng.pek2.redhat.com systemd[1]: Started rhc daemon.
Mar 12 05:11:05 dell-per6525-02.rhts.eng.pek2.redhat.com rhcd[70902]: [rhcd] 2025/03/12 05:11:05 /builddir/build/BUILD/rhc/yggdrasil-0.2.6/cmd/yggd/main.go:197: starting rhcd version 0.2.6
Mar 12 05:11:05 dell-per6525-02.rhts.eng.pek2.redhat.com rhcd[70902]: [rhcd] 2025/03/12 05:11:05 /builddir/build/BUILD/rhc/yggdrasil-0.2.6/cmd/yggd/main.go:199: attempting to kill any orphaned workers
Mar 12 05:11:05 dell-per6525-02.rhts.eng.pek2.redhat.com rhcd[70902]: [rhcd] 2025/03/12 05:11:05 /builddir/build/BUILD/rhc/yggdrasil-0.2.6/cmd/yggd/worker.go:210: process directory does not exist: /proc/70660
Mar 12 05:11:05 dell-per6525-02.rhts.eng.pek2.redhat.com rhcd[70902]: [rhcd] 2025/03/12 05:11:05 /builddir/build/BUILD/rhc/yggdrasil-0.2.6/cmd/yggd/worker.go:210: process directory does not exist: /proc/70661
```

Killing the process manually -

```
[root@dell-per6525-02 ~]# kill -9 70661
[root@dell-per6525-02 ~]# systemctl restart rhcd
[root@dell-per6525-02 ~]# rhc status
Connection status for dell-per6525-02.rhts.eng.pek2.redhat.com:

● Connected to Red Hat Subscription Management
● Connected to Red Hat Insights
● The rhc daemon is active

Manage your connected systems: https://red.ht/connector
[root@dell-per6525-02 ~]# systemctl status rhcd
● rhcd.service - rhc daemon
     Loaded: loaded (/usr/lib/systemd/system/rhcd.service; enabled; preset: disabled)
     Active: active (running) since Wed 2025-03-12 05:11:05 EDT; 1min 41s ago
       Docs: https://github.com/redhatinsights/yggdrasil
   Main PID: 70902 (rhcd)
      Tasks: 63 (limit: 1643721)
     Memory: 65.7M
        CPU: 315ms
     CGroup: /system.slice/rhcd.service
             ├─70902 /usr/sbin/rhcd
             ├─70919 /usr/libexec/rhc/rhc-package-manager-worker
             └─70920 /usr/bin/python3 /usr/libexec/rhc/rhc-worker-playbook.worker

Mar 12 05:11:06 dell-per6525-02.rhts.eng.pek2.redhat.com rhcd[70902]: [rhcd] 2025/03/12 05:11:06 /builddir/build/BUILD/rhc/yggdrasil-0.>
Mar 12 05:11:06 dell-per6525-02.rhts.eng.pek2.redhat.com rhcd[70902]: [rhcd] 2025/03/12 05:11:06 /builddir/build/BUILD/rhc/yggdrasil-0.>
Mar 12 05:11:06 dell-per6525-02.rhts.eng.pek2.redhat.com rhcd[70902]: [rhcd] 2025/03/12 05:11:06 /builddir/build/BUILD/rhc/yggdrasil-0.>
Mar 12 05:11:06 dell-per6525-02.rhts.eng.pek2.redhat.com rhcd[70902]: [rhcd] 2025/03/12 05:11:06 /builddir/build/BUILD/rhc/yggdrasil-0.>
Mar 12 05:11:06 dell-per6525-02.rhts.eng.pek2.redhat.com rhcd[70902]: [rhcd] 2025/03/12 05:11:06 /builddir/build/BUILD/rhc/yggdrasil-0.>
Mar 12 05:11:06 dell-per6525-02.rhts.eng.pek2.redhat.com rhcd[70902]: [rhcd] 2025/03/12 05:11:06 /builddir/build/BUILD/rhc/yggdrasil-0.>
Mar 12 05:11:11 dell-per6525-02.rhts.eng.pek2.redhat.com rhcd[70902]: [rhcd] 2025/03/12 05:11:11 /builddir/build/BUILD/rhc/yggdrasil-0.>
Mar 12 05:11:11 dell-per6525-02.rhts.eng.pek2.redhat.com rhcd[70902]: [rhcd] 2025/03/12 05:11:11 /builddir/build/BUILD/rhc/yggdrasil-0.>
Mar 12 05:11:11 dell-per6525-02.rhts.eng.pek2.redhat.com rhcd[70902]: [rhcd] 2025/03/12 05:11:11 /builddir/build/BUILD/rhc/yggdrasil-0.>
Mar 12 05:11:12 dell-per6525-02.rhts.eng.pek2.redhat.com rhcd[70902]: [rhcd] 2025/03/12 05:11:12 /builddir/build/BUILD/rhc/yggdrasil-0.>
[root@dell-per6525-02 ~]#
```

Archana-PandeyM commented:
LGTM.

jirihnidek reviewed (approved):
LGTM
